src/librbd/internal.cc

```diff
@@ -3513,4 +3513,143 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
       ictx->perfcounter->inc(l_librbd_readahead_bytes, readahead_length);
     }
   }
+
+  // Consistency groups functions
+
+  int create_cg(librados::IoCtx& io_ctx, const char *cg_name)
+  {
+    ostringstream bid_ss;
+    uint32_t extra;
+    int remove_r;
+    string id, id_cg, header_oid;
+
+    CephContext *cct = (CephContext *)io_ctx.cct();
+
+    Rados rados(io_ctx);
+    uint64_t bid = rados.get_instance_id();
+
+    id_cg = util::id_cg_name(cg_name);
+
+    int r = io_ctx.create(id_cg, true);
+    if (r < 0) {
+      lderr(cct) << "error creating consistency group id object: "
+	<< cpp_strerror(r)
+	<< dendl;
+      return r;
+    }
+
+    extra = rand() % 0xFFFFFFFF;
+    bid_ss << std::hex << bid << std::hex << extra;
+    id = bid_ss.str();
+    r = cls_client::set_id(&io_ctx, id_cg, id);
+    if (r < 0) {
+      lderr(cct) << "error setting consistency group id: "
+	<< cpp_strerror(r)
+	<< dendl;
+      goto err_remove_id;
+    }
+
+    ldout(cct, 2) << "adding consistency group to directory..." << dendl;
+
+    r = cls_client::dir_add_cg(&io_ctx, CG_DIRECTORY, cg_name, id);
+    if (r < 0) {
+      lderr(cct) << "error adding consistency group to directory: "
+	<< cpp_strerror(r)
+	<< dendl;
+      goto err_remove_id;
+    }
+    header_oid = util::cg_header_name(id);
+
+    r = cls_client::create_cg(&io_ctx, header_oid);
+    if (r < 0) {
+      lderr(cct) << "error writing header: " << cpp_strerror(r) << dendl;
+      goto err_remove_from_dir;
+    }
+
+    return 0;
+
+  err_remove_from_dir:
+    remove_r = cls_client::dir_remove_cg(&io_ctx, CG_DIRECTORY,
+	cg_name, id);
+    if (remove_r < 0) {
+      lderr(cct) << "error cleaning up consistency group from rbd_directory object "
+	<< "after creation failed: " << cpp_strerror(remove_r)
+	<< dendl;
+    }
+  err_remove_id:
+    remove_r = io_ctx.remove(id_cg);
+    if (remove_r < 0) {
+      lderr(cct) << "error cleaning up id object after creation failed: "
+	<< cpp_strerror(remove_r) << dendl;
+    }
+
+    return r;
+  }
+
+  int remove_cg(librados::IoCtx& io_ctx, const char *cg_name)
+  {
+    CephContext *cct((CephContext *)io_ctx.cct());
+    ldout(cct, 20) << "remove_cg " << &io_ctx << " " << cg_name << dendl;
+
+    int r = io_ctx.remove(util::id_cg_name(cg_name));
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error removing cg id object" << dendl;
+      return r;
+    }
+
+    std::string cg_id;
+    r = cls_client::dir_get_id(&io_ctx, CG_DIRECTORY,
+	                       std::string(cg_name), &cg_id);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error getting id of cg" << dendl;
+      return r;
+    }
+
+    r = cls_client::dir_remove_cg(&io_ctx, CG_DIRECTORY,
+					 cg_name, cg_id);
+    if (r < 0 && r != -ENOENT) {
+      lderr(cct) << "error removing cg from directory" << dendl;
+      return r;
+    }
+
+    string header_oid = util::cg_header_name(cg_id);
+
+    r = io_ctx.remove(header_oid);
```

Comment:
Remove header before removing from the directory so that we don't leave an object behind in case of crash.